File: src/modules/records/records.js
import { Router } from 'express';
import { z } from 'zod';
import { ApiError } from '../../lib/api-error.js';
import { date, dateTime, time, note, parse, parseListQuery, positiveId } from '../../lib/validation.js';

const configs = {
  food: {
    table: 'food_record', dateColumn: 'intake_date', view: 'v_daily_food_summary',
    create: z.object({ foodId: positiveId, mealType: z.enum(['breakfast', 'lunch', 'dinner', 'snack']), amount: z.number().positive().max(10000), amountUnit: z.enum(['g', 'ml']), intakeDate: date, note }).strict(),
    columns: { foodId: 'food_id', mealType: 'meal_type', amount: 'amount', amountUnit: 'amount_unit', intakeDate: 'intake_date', note: 'note' },
  },
  water: {
    table: 'water_record', dateColumn: 'date(intake_time)', view: 'v_daily_water_summary',
    create: z.object({ amountMl: z.number().int().min(50).max(3000), source: z.enum(['water', 'tea', 'soup', 'other']).default('water'), intakeTime: dateTime, note }).strict(),
    columns: { amountMl: 'amount_ml', source: 'source', intakeTime: 'intake_time', note: 'note' },
  },
  sport: {
    table: 'sport_record', dateColumn: 'record_date', view: 'v_daily_sport_summary',
    create: z.object({ sportType: z.string().trim().min(1).max(50), intensity: z.enum(['low', 'medium', 'high']).default('medium'), durationMin: z.number().int().min(1).max(1440), caloriesBurned: z.number().min(0).max(20000).default(0), recordDate: date, startTime: time.nullable().optional(), note }).strict(),
    columns: { sportType: 'sport_type', intensity: 'intensity', durationMin: 'duration_min', caloriesBurned: 'calories_burned', recordDate: 'record_date', startTime: 'start_time', note: 'note' },
  },
  sleep: {
    table: 'sleep_record', dateColumn: 'record_date', view: 'v_daily_sleep_summary',
    create: z.object({ sleepTime: dateTime, wakeTime: dateTime, qualityScore: z.number().int().min(0).max(100), recordDate: date, note }).strict(),
    columns: { sleepTime: 'sleep_time', wakeTime: 'wake_time', qualityScore: 'quality_score', recordDate: 'record_date', note: 'note' },
  },
};

export function createRecordsRouter(db, type) {
  const config = configs[type];
  const router = Router();
  const partial = config.create.partial();

  router.get('/daily-summary', (req, res, next) => {
    try { res.json({ data: summaries(db, type, req.auth.userId, parseListQuery(req.query)) }); } catch (error) { next(error); }
  });
  router.get('/', (req, res, next) => {
    try { res.json({ data: list(db, type, req.auth.userId, parseListQuery(req.query)) }); } catch (error) { next(error); }
  });
  router.post('/', (req, res, next) => {
    try {
      const input = parse(config.create, req.body);
      validateDomain(db, type, input);
      const keys = Object.keys(input);
      const result = db.prepare(`INSERT INTO ${config.table} (user_id, ${keys.map((key) => config.columns[key]).join(', ')}) VALUES (?, ${keys.map(() => '?').join(', ')})`).run(req.auth.userId, ...keys.map((key) => input[key]));
      res.status(201).json({ data: findOwned(db, type, result.lastInsertRowid, req.auth.userId) });
    } catch (error) { next(error); }
  });
  router.get('/:recordId', (req, res, next) => {
    try { res.json({ data: findOwned(db, type, parse(positiveId, req.params.recordId), req.auth.userId) }); } catch (error) { next(error); }
  });
  router.patch('/:recordId', (req, res, next) => {
    try {
      const id = parse(positiveId, req.params.recordId); const input = parse(partial, req.body);
      if (!Object.keys(input).length) throw new ApiError(400, 'VALIDATION_ERROR', 'Request body cannot be empty.');
      const current = findOwned(db, type, id, req.auth.userId);
      const merged = { ...toInput(type, current), ...input };
      validateDomain(db, type, merged);
      const keys = Object.keys(input);
      db.prepare(`UPDATE ${config.table} SET ${keys.map((key) => `${config.columns[key]} = ?`).join(', ')} WHERE record_id = ? AND user_id = ?`).run(...keys.map((key) => input[key]), id, req.auth.userId);
      res.json({ data: findOwned(db, type, id, req.auth.userId) });
    } catch (error) { next(error); }
  });
  router.delete('/:recordId', (req, res, next) => {
    try { const result = db.prepare(`DELETE FROM ${config.table} WHERE record_id = ? AND user_id = ?`).run(parse(positiveId, req.params.recordId), req.auth.userId); if (!result.changes) throw new ApiError(404, 'RESOURCE_NOT_FOUND', 'Record was not found.'); res.status(204).end(); } catch (error) { next(error); }
  });
  return router;
}

function list(db, type, userId, query) {
  const c = configs[type]; const clauses = ['r.user_id = ?']; const values = [userId];
  if (query.date) { clauses.push(`${type === 'water' ? 'date(r.intake_time)' : `r.${c.dateColumn}`} = ?`); values.push(query.date); }
  if (query.from) { clauses.push(`${type === 'water' ? 'date(r.intake_time)' : `r.${c.dateColumn}`} BETWEEN ? AND ?`); values.push(query.from, query.to); }
  const where = clauses.join(' AND ');
  const join = type === 'food' ? ' JOIN food_library f ON f.food_id = r.food_id' : '';
  const total = db.prepare(`SELECT COUNT(*) AS count FROM ${c.table} r${join} WHERE ${where}`).get(...values).count;
  const rows = db.prepare(`SELECT r.*${type === 'food' ? ', f.name AS food_name, f.calorie_kcal, f.protein_g, f.water_ml, f.unit_basis' : ''} FROM ${c.table} r${join} WHERE ${where} ORDER BY ${type === 'water' ? 'r.intake_time' : 'r.record_date'} DESC, r.record_id DESC LIMIT ? OFFSET ?`).all(...values, query.pageSize, (query.page - 1) * query.pageSize);
  return { items: rows.map((row) => map(type, row)), page: query.page, pageSize: query.pageSize, total };
}
function findOwned(db, type, id, userId) {
  const join = type === 'food' ? ' JOIN food_library f ON f.food_id = r.food_id' : '';
  const row = db.prepare(`SELECT r.*${type === 'food' ? ', f.name AS food_name, f.calorie_kcal, f.protein_g, f.water_ml, f.unit_basis' : ''} FROM ${configs[type].table} r${join} WHERE r.record_id = ? AND r.user_id = ?`).get(id, userId);
  if (!row) throw new ApiError(404, 'RESOURCE_NOT_FOUND', 'Record was not found.'); return map(type, row);
}
function summaries(db, type, userId, query) {
  const c = configs[type]; const dateField = type === 'food' ? 'intake_date' : type === 'water' ? 'stat_date' : 'record_date';
  const clauses = ['user_id = ?']; const values = [userId]; if (query.date) { clauses.push(`${dateField} = ?`); values.push(query.date); } if (query.from) { clauses.push(`${dateField} BETWEEN ? AND ?`); values.push(query.from, query.to); }
  const rows = db.prepare(`SELECT * FROM ${c.view} WHERE ${clauses.join(' AND ')} ORDER BY ${dateField}`).all(...values);
  return rows.map((row) => mapSummary(type, row));
}
function validateDomain(db, type, input) {
  if (type === 'sleep') { if (input.wakeTime <= input.sleepTime) throw new ApiError(400, 'VALIDATION_ERROR', 'wakeTime must be later than sleepTime.'); if (input.recordDate !== input.wakeTime.slice(0, 10)) throw new ApiError(400, 'VALIDATION_ERROR', 'recordDate must match wakeTime date.'); }
  if (type === 'food') { const food = db.prepare('SELECT unit_basis FROM food_library WHERE food_id = ?').get(input.foodId); if (!food) throw new ApiError(400, 'VALIDATION_ERROR', 'Food does not exist.'); const expected = food.unit_basis === 'per_100g' ? 'g' : 'ml'; if (input.amountUnit !== expected) throw new ApiError(400, 'VALIDATION_ERROR', `Food amountUnit must be ${expected}.`); }
}
function map(type, r) {
  const base = { id: r.record_id, note: r.note };
  if (type === 'food') return { ...base, foodId:r.food_id, mealType:r.meal_type, amount:r.amount, amountUnit:r.amount_unit, intakeDate:r.intake_date, food:{id:r.food_id,name:r.food_name,calorieKcal:r.calorie_kcal,proteinG:r.protein_g,waterMl:r.water_ml,unitBasis:r.unit_basis} };
  if (type === 'water') return { ...base, amountMl:r.amount_ml, source:r.source, intakeTime:r.intake_time };
  if (type === 'sport') return { ...base, sportType:r.sport_type,intensity:r.intensity,durationMin:r.duration_min,caloriesBurned:r.calories_burned,recordDate:r.record_date,startTime:r.start_time };
  return { ...base, sleepTime:r.sleep_time,wakeTime:r.wake_time,qualityScore:r.quality_score,recordDate:r.record_date };
}
function toInput(type, r) { const { id, ...input } = map(type, r); if (type === 'food') delete input.food; return input; }
function mapSummary(type, r) { if(type==='food')return{date:r.intake_date,totalCalorieKcal:r.total_calorie_kcal,totalProteinG:r.total_protein_g,totalFoodWaterMl:r.total_food_water_ml,foodHealthScore:r.food_health_score}; if(type==='water')return{date:r.stat_date,totalWaterMl:r.total_water_ml,drinkTimes:r.drink_times}; if(type==='sport')return{date:r.record_date,totalDurationMin:r.total_duration_min,totalCaloriesBurned:r.total_calories_burned,sportTimes:r.sport_times}; return{date:r.record_date,sleepHours:r.sleep_hours,sleepQualityScore:r.sleep_quality_score}; }

File: backend-member3/app/routers/records.py
import sqlite3
from dataclasses import dataclass
from typing import Any, Type

from fastapi import APIRouter, Depends
from pydantic import BaseModel, create_model

from ..dependencies import make_dependencies
from ..errors import ApiError
from ..schemas.common import ListQuery, require_nonempty
from ..schemas.records import FoodRecordInput, SleepRecordInput, SportRecordInput, WaterRecordInput


@dataclass(frozen=True)
class RecordSpec:
    table: str
    view: str
    date_column: str
    summary_date: str
    input_model: Type[BaseModel]
    columns: dict[str, str]


SPECS = {
    "food": RecordSpec("food_record", "v_daily_food_summary", "intake_date", "intake_date", FoodRecordInput, {"food_id": "food_id", "meal_type": "meal_type", "amount": "amount", "amount_unit": "amount_unit", "intake_date": "intake_date", "note": "note"}),
    "water": RecordSpec("water_record", "v_daily_water_summary", "date(intake_time)", "stat_date", WaterRecordInput, {"amount_ml": "amount_ml", "source": "source", "intake_time": "intake_time", "note": "note"}),
    "sport": RecordSpec("sport_record", "v_daily_sport_summary", "record_date", "record_date", SportRecordInput, {"sport_type": "sport_type", "intensity": "intensity", "duration_min": "duration_min", "calories_burned": "calories_burned", "record_date": "record_date", "start_time": "start_time", "note": "note"}),
    "sleep": RecordSpec("sleep_record", "v_daily_sleep_summary", "record_date", "record_date", SleepRecordInput, {"sleep_time": "sleep_time", "wake_time": "wake_time", "quality_score": "quality_score", "record_date": "record_date", "note": "note"}),
}


def patch_model(model: Type[BaseModel]) -> Type[BaseModel]:
    fields = {name: (field.annotation | None, None) for name, field in model.model_fields.items()}
    return create_model(f"{model.__name__}Patch", __base__=model, **fields)


def map_record(kind: str, row: sqlite3.Row) -> dict[str, Any]:
    base = {"id": row["record_id"], "note": row["note"]}
    if kind == "food":
        return {**base, "foodId": row["food_id"], "mealType": row["meal_type"], "amount": row["amount"], "amountUnit": row["amount_unit"], "intakeDate": row["intake_date"], "food": {"id": row["food_id"], "name": row["food_name"], "calorieKcal": row["calorie_kcal"], "proteinG": row["protein_g"], "waterMl": row["water_ml"], "unitBasis": row["unit_basis"]}}
    if kind == "water":
        return {**base, "amountMl": row["amount_ml"], "source": row["source"], "intakeTime": row["intake_time"]}
    if kind == "sport":
        return {**base, "sportType": row["sport_type"], "intensity": row["intensity"], "durationMin": row["duration_min"], "caloriesBurned": row["calories_burned"], "recordDate": row["record_date"], "startTime": row["start_time"]}
    return {**base, "sleepTime": row["sleep_time"], "wakeTime": row["wake_time"], "qualityScore": row["quality_score"], "recordDate": row["record_date"]}


def map_summary(kind: str, row: sqlite3.Row) -> dict[str, Any]:
    if kind == "food":
        return {"date": row["intake_date"], "totalCalorieKcal": row["total_calorie_kcal"], "totalProteinG": row["total_protein_g"], "totalFoodWaterMl": row["total_food_water_ml"], "foodHealthScore": row["food_health_score"]}
    if kind == "water":
        return {"date": row["stat_date"], "totalWaterMl": row["total_water_ml"], "drinkTimes": row["drink_times"]}
    if kind == "sport":
        return {"date": row["record_date"], "totalDurationMin": row["total_duration_min"], "totalCaloriesBurned": row["total_calories_burned"], "sportTimes": row["sport_times"]}
    return {"date": row["record_date"], "sleepHours": row["sleep_hours"], "sleepQualityScore": row["sleep_quality_score"]}


def row_query(kind: str) -> str:
    spec = SPECS[kind]
    extra = ", f.name AS food_name, f.calorie_kcal, f.protein_g, f.water_ml, f.unit_basis" if kind == "food" else ""
    join = " JOIN food_library f ON f.food_id = r.food_id" if kind == "food" else ""
    return f"SELECT r.*{extra} FROM {spec.table} r{join}"


def find_owned(connection: sqlite3.Connection, kind: str, record_id: int, user_id: int) -> dict:
    row = connection.execute(f"{row_query(kind)} WHERE r.record_id = ? AND r.user_id = ?", (record_id, user_id)).fetchone()
    if not row:
        raise ApiError(404, "RESOURCE_NOT_FOUND", "Record was not found.")
    return map_record(kind, row)


def validate_domain(connection: sqlite3.Connection, kind: str, values: dict[str, Any]) -> None:
    if kind == "sleep":
        if values["wake_time"] <= values["sleep_time"]:
            raise ApiError(400, "VALIDATION_ERROR", "wakeTime must be later than sleepTime.")
        if values["record_date"] != values["wake_time"][:10]:
            raise ApiError(400, "VALIDATION_ERROR", "recordDate must match wakeTime date.")
    if kind == "food":
        food = connection.execute("SELECT unit_basis FROM food_library WHERE food_id = ?", (values["food_id"],)).fetchone()
        if not food:
            raise ApiError(400, "VALIDATION_ERROR", "Food does not exist.")
        expected = "g" if food["unit_basis"] == "per_100g" else "ml"
        if values["amount_unit"] != expected:
            raise ApiError(400, "VALIDATION_ERROR", f"Food amountUnit must be {expected}.")


def router(settings) -> APIRouter:
    routes = APIRouter(tags=["records"])
    get_db, principal = make_dependencies(settings)

    for kind, spec in SPECS.items():
        path = f"/{kind}-records"
        create = spec.input_model
        patch = patch_model(create)

        def summaries(query: ListQuery = Depends(), user=Depends(principal), connection: sqlite3.Connection = Depends(get_db), _kind=kind, _spec=spec):
            query.validate_range()
            clauses, values = ["user_id = ?"], [user["user_id"]]
            if query.date:
                clauses.append(f"{_spec.summary_date} = ?")
                values.append(query.date)
            elif query.from_date:
                clauses.append(f"{_spec.summary_date} BETWEEN ? AND ?")
                values.extend([query.from_date, query.to])
            rows = connection.execute(f"SELECT * FROM {_spec.view} WHERE {' AND '.join(clauses)} ORDER BY {_spec.summary_date}", values).fetchall()
            return {"data": [map_summary(_kind, row) for row in rows]}

        def list_records(query: ListQuery = Depends(), user=Depends(principal), connection: sqlite3.Connection = Depends(get_db), _kind=kind, _spec=spec):
            query.validate_range()
            clauses, values = ["r.user_id = ?"], [user["user_id"]]
            field = _spec.date_column if _kind != "water" else "date(r.intake_time)"
            if query.date:
                clauses.append(f"{field} = ?")
                values.append(query.date)
            elif query.from_date:
                clauses.append(f"{field} BETWEEN ? AND ?")
                values.extend([query.from_date, query.to])
            where = " AND ".join(clauses)
            join = " JOIN food_library f ON f.food_id = r.food_id" if _kind == "food" else ""
            total = connection.execute(f"SELECT COUNT(*) AS count FROM {_spec.table} r{join} WHERE {where}", values).fetchone()["count"]
            order = "r.intake_time" if _kind == "water" else f"r.{_spec.date_column}"
            rows = connection.execute(f"{row_query(_kind)} WHERE {where} ORDER BY {order} DESC, r.record_id DESC LIMIT ? OFFSET ?", [*values, query.page_size, (query.page - 1) * query.page_size]).fetchall()
            return {"data": {"items": [map_record(_kind, row) for row in rows], "page": query.page, "pageSize": query.page_size, "total": total}}

        def create_record(input: create, user=Depends(principal), connection: sqlite3.Connection = Depends(get_db), _kind=kind, _spec=spec):
            values = input.model_dump()
            validate_domain(connection, _kind, values)
            keys = list(values)
            result = connection.execute(f"INSERT INTO {_spec.table} (user_id, {', '.join(_spec.columns[key] for key in keys)}) VALUES (?, {', '.join('?' for _ in keys)})", (user["user_id"], *(values[key] for key in keys)))
            return {"data": find_owned(connection, _kind, result.lastrowid, user["user_id"])}

        def detail(record_id: int, user=Depends(principal), connection: sqlite3.Connection = Depends(get_db), _kind=kind):
            if record_id <= 0:
                raise ApiError(400, "VALIDATION_ERROR", "Request validation failed.")
            return {"data": find_owned(connection, _kind, record_id, user["user_id"])}

        def update_record(record_id: int, input: patch, user=Depends(principal), connection: sqlite3.Connection = Depends(get_db), _kind=kind, _spec=spec):
            if record_id <= 0:
                raise ApiError(400, "VALIDATION_ERROR", "Request validation failed.")
            require_nonempty(input)
            current = find_owned(connection, _kind, record_id, user["user_id"])
            current_values = {key: current.get(alias) for key, alias in {"food_id": "foodId", "meal_type": "mealType", "amount": "amount", "amount_unit": "amountUnit", "intake_date": "intakeDate", "amount_ml": "amountMl", "source": "source", "intake_time": "intakeTime", "sport_type": "sportType", "intensity": "intensity", "duration_min": "durationMin", "calories_burned": "caloriesBurned", "record_date": "recordDate", "start_time": "startTime", "sleep_time": "sleepTime", "wake_time": "wakeTime", "quality_score": "qualityScore", "note": "note"}.items() if key in _spec.columns}
            changes = input.model_dump(exclude_unset=True)
            validate_domain(connection, _kind, {**current_values, **changes})
            assignments = ", ".join(f"{_spec.columns[key]} = ?" for key in changes)
            connection.execute(f"UPDATE {_spec.table} SET {assignments} WHERE record_id = ? AND user_id = ?", (*changes.values(), record_id, user["user_id"]))
            return {"data": find_owned(connection, _kind, record_id, user["user_id"])}

        def delete_record(record_id: int, user=Depends(principal), connection: sqlite3.Connection = Depends(get_db), _spec=spec):
            if record_id <= 0:
                raise ApiError(400, "VALIDATION_ERROR", "Request validation failed.")
            result = connection.execute(f"DELETE FROM {_spec.table} WHERE record_id = ? AND user_id = ?", (record_id, user["user_id"]))
            if not result.rowcount:
                raise ApiError(404, "RESOURCE_NOT_FOUND", "Record was not found.")
            return None

        routes.add_api_route(f"{path}/daily-summary", summaries, methods=["GET"])
        routes.add_api_route(path, list_records, methods=["GET"])
        routes.add_api_route(path, create_record, methods=["POST"], status_code=201)
        routes.add_api_route(f"{path}/{{record_id}}", detail, methods=["GET"])
        routes.add_api_route(f"{path}/{{record_id}}", update_record, methods=["PATCH"])
        routes.add_api_route(f"{path}/{{record_id}}", delete_record, methods=["DELETE"], status_code=204)

    return routes
